File: paseto-core/src/main/java/net/aholbrook/paseto/PasetoV2.java
package net.aholbrook.paseto;

import net.aholbrook.paseto.base64.jvm8.Base64Provider;
import net.aholbrook.paseto.crypto.Pair;
import net.aholbrook.paseto.keys.KeyPair;
import net.aholbrook.paseto.crypto.NonceGenerator;
import net.aholbrook.paseto.crypto.v2.V2CryptoLoader;
import net.aholbrook.paseto.crypto.v2.V2CryptoProvider;
import net.aholbrook.paseto.encoding.EncodingProvider;
import net.aholbrook.paseto.exception.DecryptionException;
import net.aholbrook.paseto.exception.PasetoParseException;
import net.aholbrook.paseto.exception.SignatureVerificationException;
import net.aholbrook.paseto.keys.AsymmetricPublicKey;
import net.aholbrook.paseto.keys.AsymmetricSecretKey;
import net.aholbrook.paseto.keys.SymmetricKey;
import net.aholbrook.paseto.util.PaeUtil;
import net.aholbrook.paseto.util.StringUtils;

public class PasetoV2 extends Paseto {
	private final static String VERSION = "v2";
	public final static String HEADER_LOCAL = VERSION + SEPARATOR + PURPOSE_LOCAL + SEPARATOR; // v2.local.
	public final static String HEADER_PUBLIC = VERSION + SEPARATOR + PURPOSE_PUBLIC + SEPARATOR; // v2.public.

	private final V2CryptoProvider cryptoProvider;

	private PasetoV2(Base64Provider base64Provider, EncodingProvider encodingProvider, V2CryptoProvider cryptoProvider,
			NonceGenerator nonceGenerator) {
		super(base64Provider, encodingProvider, nonceGenerator);
		this.cryptoProvider = cryptoProvider;
	}

	@Override
	public String encrypt(Object payload, SymmetricKey key, String footer) {
		// Verify key version.
		key.verifyKey(Version.V2);

		footer = StringUtils.ntes(footer); // convert null to ""
		byte[] payloadBytes = StringUtils.getBytesUtf8(encodingProvider.encode(payload));
		byte[] footerBytes = StringUtils.getBytesUtf8(footer);

		byte nonce[] = nonceGenerator.generateNonce();
		byte[] n = new byte[cryptoProvider.xChaCha20Poly1305IetfNpubbytes()];
		cryptoProvider.blake2b(n, payloadBytes, nonce);

		byte[] preAuth = PaeUtil.pae(StringUtils.getBytesUtf8(HEADER_LOCAL), n, footerBytes);

		byte[] c = new byte[payloadBytes.length + cryptoProvider.xChaCha20Poly1305IetfAbytes()];
		cryptoProvider.aeadXChaCha20Poly1305IetfEncrypt(c, payloadBytes, preAuth, n, key.getMaterial());

		byte[] nc = new byte[n.length + c.length];
		System.arraycopy(n, 0, nc, 0, n.length);
		System.arraycopy(c, 0, nc, n.length, c.length);

		if (footerBytes.length > 0) {
			return HEADER_LOCAL + base64Provider.encodeToString(nc) + SEPARATOR
					+ base64Provider.encodeToString(footerBytes);
		} else {
			return HEADER_LOCAL + base64Provider.encodeToString(nc);
		}
	}

	@Override
	public <_Payload> _Payload decrypt(String token, SymmetricKey key, String footer, Class<_Payload> payloadClass) {
		// Verify key version.
		key.verifyKey(Version.V2);

		// Split token into sections
		String[] sections = split(token);
		if (sections == null) {
			throw new PasetoParseException(PasetoParseException.Reason.MISSING_SECTIONS, token);
		}

		// Check header
		checkHeader(token, sections, HEADER_LOCAL);

		// Decode footer
		String decodedFooter = decodeFooter(token, sections, footer);

		// Decrypt
		byte[] nc = base64Provider.decodeFromString(sections[2]);
		byte[] n = new byte[cryptoProvider.xChaCha20Poly1305IetfNpubbytes()];
		// verify length
		if (nc.length < n.length + 1) {
			throw new PasetoParseException(PasetoParseException.Reason.PAYLOAD_LENGTH, token)
					.setMinLength(n.length + 1);
		}
		byte[] c = new byte[nc.length - n.length];
		System.arraycopy(nc, 0, n, 0, n.length);
		System.arraycopy(nc, n.length, c, 0, c.length);

		byte[] preAuth = PaeUtil.pae(StringUtils.getBytesUtf8(HEADER_LOCAL), n, StringUtils.getBytesUtf8(decodedFooter));
		byte[] p = new byte[c.length - cryptoProvider.xChaCha20Poly1305IetfAbytes()];
		if (!cryptoProvider.aeadXChaCha20Poly1305IetfDecrypt(p, c, preAuth, n, key.getMaterial())) {
			throw new DecryptionException(token);
		}

		// Convert from JSON
		return decode(p, payloadClass);
	}

	@Override
	public String sign(Object payload, AsymmetricSecretKey sk, String footer) {
		// Verify key version.
		sk.verifyKey(Version.V2);

		footer = StringUtils.ntes(footer); // convert null to ""
		byte[] payloadBytes = StringUtils.getBytesUtf8(encodingProvider.encode(payload));
		byte[] footerBytes = StringUtils.getBytesUtf8(footer);

		byte[] m2 = PaeUtil.pae(StringUtils.getBytesUtf8(HEADER_PUBLIC), payloadBytes, footerBytes);
		byte[] sig = new byte[cryptoProvider.ed25519SignBytes()];
		cryptoProvider.ed25519Sign(sig, m2, sk.getMaterial());

		byte[] msig = new byte[payloadBytes.length + sig.length];
		System.arraycopy(payloadBytes, 0, msig, 0, payloadBytes.length);
		System.arraycopy(sig, 0, msig, payloadBytes.length, sig.length);

		if (footerBytes.length > 0) {
			return HEADER_PUBLIC + base64Provider.encodeToString(msig)
					+ SEPARATOR + base64Provider.encodeToString(footerBytes);
		} else {
			return HEADER_PUBLIC + base64Provider.encodeToString(msig);
		}
	}

	@Override
	public <_Payload> _Payload verify(String token, AsymmetricPublicKey pk, String footer, Class<_Payload> payloadClass) {
		// Verify key version.
		pk.verifyKey(Version.V2);

		// Split token into sections
		String[] sections = split(token);
		if (sections == null) {
			throw new PasetoParseException(PasetoParseException.Reason.MISSING_SECTIONS, token);
		}

		// Check header
		checkHeader(token, sections, HEADER_PUBLIC);

		// Decode footer
		String decodedFooter = decodeFooter(token, sections, footer);

		// Verify
		byte[] msig = base64Provider.decodeFromString(sections[2]);
		byte[] s = new byte[cryptoProvider.ed25519SignBytes()];
		// verify length
		if (msig.length < s.length + 1) {
			throw new PasetoParseException(PasetoParseException.Reason.PAYLOAD_LENGTH, token)
					.setMinLength(s.length + 1);
		}
		byte[] m = new byte[msig.length - s.length];
		System.arraycopy(msig, msig.length - s.length, s, 0, s.length);
		System.arraycopy(msig, 0, m, 0, m.length);

		byte[] m2 = PaeUtil.pae(StringUtils.getBytesUtf8(HEADER_PUBLIC), m, StringUtils.getBytesUtf8(decodedFooter));
		if (!cryptoProvider.ed25519Verify(s, m2, pk.getMaterial())) {
			throw new SignatureVerificationException(token);
		}

		// Convert from JSON
		return decode(m, payloadClass);
	}

	@Override
	public KeyPair generateKeyPair() {
		Pair<byte[], byte[]> rawKey = cryptoProvider.ed25519Generate();
		return new KeyPair(
				new AsymmetricSecretKey(rawKey.a, Version.V2),
				new AsymmetricPublicKey(rawKey.b, Version.V2)
		);
	}

	public static class Builder extends Paseto.Builder {
		V2CryptoProvider v2CryptoProvider;

		public Builder withV2CryptoProvider(V2CryptoProvider v2CryptoProvider) {
			this.v2CryptoProvider = v2CryptoProvider;
			return this;
		}

		@Override
		protected void fillInDefaults() {
			super.fillInDefaults();
			if (v2CryptoProvider == null) { v2CryptoProvider = V2CryptoLoader.getProvider(); }
			if (nonceGenerator == null) { nonceGenerator = v2CryptoProvider.getNonceGenerator(); }
		}

		public PasetoV2 build() {
			fillInDefaults();
			return new PasetoV2(base64Provider, encodingProvider, v2CryptoProvider, nonceGenerator);
		}
	}
}

File: paseto/src/main/kotlin/net/aholbrook/paseto/protocol/PasetoV2.kt
@file:Suppress("MagicNumber")

package net.aholbrook.paseto.protocol

import net.aholbrook.paseto.UrlSafeNoPadding
import net.aholbrook.paseto.crypto.ED25519_BYTES
import net.aholbrook.paseto.crypto.XCHACHA20_POLY1305_IETF_ABYTES
import net.aholbrook.paseto.crypto.XCHACHA20_POLY1305_IETF_NPUBBYTES
import net.aholbrook.paseto.crypto.aeadXChaCha20Poly1305IetfDecrypt
import net.aholbrook.paseto.crypto.aeadXChaCha20Poly1305IetfEncrypt
import net.aholbrook.paseto.crypto.blake2b
import net.aholbrook.paseto.crypto.ed25519Sign
import net.aholbrook.paseto.crypto.ed25519Verify
import net.aholbrook.paseto.crypto.generateNonce
import net.aholbrook.paseto.crypto.pae
import net.aholbrook.paseto.decodeOrNull
import net.aholbrook.paseto.exception.DecryptionException
import net.aholbrook.paseto.exception.EncryptionException
import net.aholbrook.paseto.exception.InvalidHeaderException
import net.aholbrook.paseto.exception.PasetoParseException
import net.aholbrook.paseto.exception.SignatureVerificationException
import net.aholbrook.paseto.exception.SigningException
import net.aholbrook.paseto.protocol.key.AsymmetricPublicKey
import net.aholbrook.paseto.protocol.key.AsymmetricSecretKey
import net.aholbrook.paseto.protocol.key.SymmetricKey
import kotlin.io.encoding.Base64

internal object PasetoV2 : Paseto {
    override val version: Version = Version.V2
    override val supportsImplicitAssertion: Boolean = false

    override fun encrypt(m: ByteArray, key: SymmetricKey, footer: String, implicitAssertion: String): String {
        val cleanup = mutableListOf<Runnable>()

        try {
            val k = key.getKeyMaterialFor(Version.V2, Purpose.LOCAL)
            val h = "v2.local."
            val f = footer.toByteArray(Charsets.UTF_8)
            val b: ByteArray = generateNonce(24)
            cleanup.add { b.fill(0) }
            val n = ByteArray(XCHACHA20_POLY1305_IETF_NPUBBYTES)
            blake2b(n, b, m)

            val preAuth = pae(h.toByteArray(Charsets.UTF_8), n, f)
            val c = ByteArray(m.size + XCHACHA20_POLY1305_IETF_ABYTES)
            if (!aeadXChaCha20Poly1305IetfEncrypt(c, m, preAuth, n, k)) {
                throw EncryptionException()
            }

            return h + Base64.UrlSafeNoPadding.encode(n + c) +
                if (f.isEmpty()) {
                    ""
                } else {
                    ".${Base64.UrlSafeNoPadding.encode(f)}"
                }
        } finally {
            key.clear()
            cleanup.forEach { it.run() }
        }
    }

    override fun decrypt(
        token: String,
        key: SymmetricKey,
        footer: String?,
        implicitAssertion: String,
    ): Pair<String, String> {
        try {
            val k = key.getKeyMaterialFor(Version.V2, Purpose.LOCAL)
            val h = "v2.local."
            val sections = split(token)
            val f = decodeFooter(token, sections, footer)

            // Check header
            if (!token.startsWith(h)) {
                throw InvalidHeaderException(sections.version + SEPARATOR + sections.purpose + SEPARATOR, h, token)
            }

            // Decrypt
            val nc = Base64.UrlSafeNoPadding.decodeOrNull(sections.payload)
                ?: throw PasetoParseException(PasetoParseException.Reason.INVALID_BASE64, token)
            // verify length
            if (nc.size < XCHACHA20_POLY1305_IETF_NPUBBYTES + 1) {
                throw PasetoParseException(PasetoParseException.Reason.PAYLOAD_LENGTH, token).apply {
                    minLength = XCHACHA20_POLY1305_IETF_NPUBBYTES + 1
                }
            }
            val n = nc.copyOfRange(0, XCHACHA20_POLY1305_IETF_NPUBBYTES)
            val c = nc.copyOfRange(XCHACHA20_POLY1305_IETF_NPUBBYTES, nc.size)

            val preAuth = pae(h.toByteArray(Charsets.UTF_8), n, f.toByteArray(Charsets.UTF_8))
            val p = ByteArray(c.size - XCHACHA20_POLY1305_IETF_ABYTES)
            if (!aeadXChaCha20Poly1305IetfDecrypt(p, c, preAuth, n, k)) {
                throw DecryptionException(token)
            }

            // Convert from JSON
            return Pair(p.toString(Charsets.UTF_8), f)
        } finally {
            key.clear()
        }
    }

    override fun sign(
        m: ByteArray,
        secretKey: AsymmetricSecretKey,
        footer: String,
        implicitAssertion: String,
    ): String {
        try {
            // Verify key version.
            val k = secretKey.getKeyMaterialFor(Version.V2, Purpose.PUBLIC)
            val h = "v2.public."
            val f = footer.toByteArray(Charsets.UTF_8)

            val m2 = pae(h.toByteArray(Charsets.UTF_8), m, f)
            val sig = ByteArray(ED25519_BYTES)
            if (!ed25519Sign(sig, m2, k)) {
                throw SigningException(m)
            }

            return h + Base64.UrlSafeNoPadding.encode(m + sig) +
                if (f.isEmpty()) {
                    ""
                } else {
                    ".${Base64.UrlSafeNoPadding.encode(f)}"
                }
        } finally {
            secretKey.clear()
        }
    }

    override fun verify(
        token: String,
        publicKey: AsymmetricPublicKey,
        footer: String?,
        implicitAssertion: String,
    ): Pair<String, String> {
        val k = publicKey.getKeyMaterialFor(Version.V2, Purpose.PUBLIC)
        val h = "v2.public."
        val sections = split(token)
        val f = decodeFooter(token, sections, footer)

        // Check header
        if (!token.startsWith(h)) {
            throw InvalidHeaderException(sections.version + SEPARATOR + sections.purpose + SEPARATOR, h, token)
        }

        // Verify
        val sm = Base64.UrlSafeNoPadding.decodeOrNull(sections.payload)
            ?: throw PasetoParseException(PasetoParseException.Reason.INVALID_BASE64, token)
        // verify length
        if (sm.size < ED25519_BYTES + 1) {
            throw PasetoParseException(PasetoParseException.Reason.PAYLOAD_LENGTH, token).apply {
                minLength = ED25519_BYTES + 1
            }
        }
        val s = sm.copyOfRange(sm.size - ED25519_BYTES, sm.size)
        val m = sm.copyOfRange(0, sm.size - ED25519_BYTES)

        val m2 = pae(h.toByteArray(Charsets.UTF_8), m, f.toByteArray(Charsets.UTF_8))
        if (!ed25519Verify(s, m2, k)) {
            throw SignatureVerificationException(token)
        }

        // Convert from JSON
        return Pair(m.toString(Charsets.UTF_8), f)
    }
}
